clippy_lints/src/types.rs

```diff
@@ -66,6 +66,34 @@ declare_clippy_lint! {
     "usage of `Box<Vec<T>>`, vector elements are already on the heap"
 }
 
+declare_clippy_lint! {
+    /// **What it does:** Checks for use of `&[char]` anywhere in the code.
+    ///
+    /// **Why is this bad?** `char` represents a Unicode codepoint, not a
+    /// character. Usually users of `&[char]` want O(1) indexing on characters,
+    /// not codepoints.
+    ///
+    /// **Known problems:** None.
```

Comment:
When dealing with FFI to other high-level languages where strings are represented by sequences of `char`s, the single correct way *is* to use `&[char]` or `Vec<char>`. This should be listed under "Known problems".

Reply:
as far as we understand it, `char` should not be used at FFI boundaries. consider `u32`/`i32` instead.

(from what we understand `&` and `&mut` shouldn't be used at FFI boundaries. basically any type that is rust ABI shouldn't be used at FFI boundaries?)

Reply:
I found another counter-example (though nightly only): [std::str::pattern::Pattern](https://doc.rust-lang.org/std/str/pattern/trait.Pattern.html) is implemented on `&[char]` (which basically means any char in the slice matches the pattern). There may be other use cases where having chars equidistantly distributed in memory may be of the essence. I feel the `Problems` section should reflect this.

Reply:
`Vec<char>` and `&[char]` is also a completely reasonable way to represent UTF32 encoded strings.